File: 프로그래머스/lv2/87946. 피로도/피로도.java
import java.util.*;
class Solution {
    boolean[] visited;
    int count;
    int answer;
    
    public int solution(int k, int[][] dungeons) {
        int len = dungeons.length;
        visited = new boolean[len];
        dfs(dungeons, k);
        
        return answer;
    }
    
    public void dfs(int[][] dungeons, int k){
        int len = dungeons.length;
        for(int i=0; i<len; i++){
            if(!visited[i] && k>=dungeons[i][0]){
                count++;
                answer=Math.max(answer, count);
                visited[i] = true;
                k -= dungeons[i][1];
                dfs(dungeons, k);
                k += dungeons[i][1];
                visited[i] = false;
                count--;
            }
        }
        
        
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    visited = [False] * len(dungeons)
    count = 0
    answer = 0
    
    def dfs(k, count):
        nonlocal answer
        for i in range(len(dungeons)):
            if not visited[i] and k >= dungeons[i][0]:
                count += 1
                answer = max(answer, count)
                visited[i] = True
                k -= dungeons[i][1]
                dfs(k, count)
                k += dungeons[i][1]
                visited[i] = False
                count -= 1
    
    dfs(k, count)
    return answer
